### PR TITLE
Implementing CooldownData

### PR DESCRIPTION
**Pull request by Rain336**
Implementation of `SpongeCooldownData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSpongeCooldownData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `CooldownValueProcessor`
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `COOLDOWN` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:
```java
@Plugin(id = "CodeTester", name = "CodeTester", version = "1.0.0")
public class CodeTester {

    @Listener
    public void onPlayerInteractBlock(InteractBlockEvent event) {
        Optional<Player> optional = event.getCause().first(Player.class);
        if (optional.isPresent()) {
            Player player = optional.get();
            if (player.getItemInHand().isPresent()) {
                ItemStack item = player.getItemInHand().get();
                Hopper hopper = (Hopper) event.getTargetBlock().getLocation().get().getTileEntity().get();
                if (item.getItem() == ItemTypes.STICK) {
                    player.sendMessage(Texts.of("HopperTransferCooldown: ", hopper.get(Keys.COOLDOWN).get()));
                }
                if (item.getItem() == ItemTypes.BLAZE_ROD) {
                    hopper.offer(Keys.COOLDOWN, 100);
                }
            }
        }
    }
}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/ImmutableSpongeCooldownData.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public CooldownData asMutable() {
+        return new SpongeCooldownData(this.getValue());
+    }
+
+    @Override
+    public ImmutableValue<Integer> cooldown() {
+        return ImmutableSpongeValue.cachedOf(Keys.COOLDOWN, -1, this.getValue());
```
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/ImmutableSpongeCooldownData.java]: Please make this a final field

**Comment by gabizou** [comment]
So, a couple of things about your test plugins:
- Your test plugin is very buggy and easily throws exceptions in many places unless you know exactly what is going on.
- You don't check any optionals, resulting in assumptions and exceptions
- You aren't testing everything as asked
  - Manipulators aren't being retrieved, you're just interacting with values directly
  - Serialization and deserialization isn't being tested.

Aside from that, I've tested it enough to know that it works, but that's only because I know the underlying code well enough, but seriously, you should consider your test code design as all corner cases should be considered.
